File: tremor-guard-backend/app/schemas/domain.py
from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class PatientProfileDTO(BaseModel):
    id: str
    name: str
    age: int
    gender: str
    diagnosis: str
    duration: str
    hospital: str
    device_id: str


class DeviceStatusDTO(BaseModel):
    battery: int
    connection: str
    connection_label: str
    last_sync: str
    available_days: str
    firmware: str


class TremorMetricSummaryDTO(BaseModel):
    label: str
    value: str | int | float
    unit: str
    subtitle: str
    tone: str


class TremorTrendPointDTO(BaseModel):
    time: str
    amplitude: float
    label: str | None = None
    medication_taken: bool = False


class AiInsightDTO(BaseModel):
    id: str
    title: str
    summary: str
    emphasis: str | None = None


class AiChatMessageInput(BaseModel):
    role: Literal["user", "assistant"]
    content: str = Field(min_length=1, max_length=4000)


class AiChatMessageResponse(BaseModel):
    role: Literal["assistant"]
    content: str


class AiChatUsage(BaseModel):
    prompt_tokens: int | None = None
    completion_tokens: int | None = None
    total_tokens: int | None = None


class AiChatRequest(BaseModel):
    messages: list[AiChatMessageInput] = Field(min_length=1, max_length=20)


class AiChatResponse(BaseModel):
    message: AiChatMessageResponse
    model: str
    usage: AiChatUsage | None = None


class MedicationEntryDTO(BaseModel):
    id: int
    time: str
    name: str
    dose: str
    status: str


class ReportSummaryDTO(BaseModel):
    id: str
    date: date
    type: str
    size: str
    status: str


class ConsentSettingsDTO(BaseModel):
    share_with_doctor: bool
    rag_analysis_enabled: bool
    cloud_sync_enabled: bool


class MeProfileResponse(BaseModel):
    patient_profile: PatientProfileDTO
    device_status: DeviceStatusDTO
    consent_settings: ConsentSettingsDTO


class PatientProfileUpsertRequest(BaseModel):
    name: str
    age: int
    gender: str
    diagnosis: str
    duration: str
    hospital: str


class ProfileCompletionStatus(BaseModel):
    onboarding_state: str
    has_profile: bool
    has_bound_device: bool


class PatientProfileUpsertResponse(BaseModel):
    patient_profile: PatientProfileDTO
    completion: ProfileCompletionStatus


class DeviceBindingDTO(BaseModel):
    id: str
    device_serial: str
    device_name: str
    firmware_version: str
    binding_status: str
    bound_at: datetime | None = None
    unbound_at: datetime | None = None


class DeviceBindingRequest(BaseModel):
    device_serial: str
    activation_code: str


class DeviceBindingResponse(BaseModel):
    device_binding: DeviceBindingDTO | None
    completion: ProfileCompletionStatus


class BindingConflictError(BaseModel):
    code: str
    detail: str


class DashboardOverviewResponse(BaseModel):
    metric_summaries: list[TremorMetricSummaryDTO]
    device_status: DeviceStatusDTO
    trend_points: list[TremorTrendPointDTO]
    overview_insight: AiInsightDTO


class MedicationListResponse(BaseModel):
    medication_entries: list[MedicationEntryDTO]


class CreateMedicationRequest(BaseModel):
    taken_at: datetime
    name: str
    dose: str
    status: str


class UpdateMedicationRequest(BaseModel):
    taken_at: datetime | None = None
    name: str | None = None
    dose: str | None = None
    status: str | None = None


class ReportListResponse(BaseModel):
    report_summaries: list[ReportSummaryDTO]


class CreateReportRequest(BaseModel):
    report_date: date | None = None
    report_type: str = "周度病情评估摘要"


class CreateReportResponse(BaseModel):
    report: ReportSummaryDTO


class TremorEventIngestItem(BaseModel):
    start_at: datetime
    duration_sec: int
    dominant_hz: float
    rms_amplitude: float
    confidence: float
    source: str


class TremorEventIngestRequest(BaseModel):
    items: list[TremorEventIngestItem]


class TremorIngestResponse(BaseModel):
    accepted_count: int
    duplicate: bool
    batch_key: str | None


AsyncJobStatus = Literal["queued", "processing", "succeeded", "failed"]
RehabPlanStatus = Literal[
    "active_only",
    "candidate_pending_confirmation",
    "candidate_confirmed",
    "candidate_superseded",
]
RehabConflictStatus = Literal["consistent", "conflicting", "insufficient_data"]
RehabGenerationEligibility = Literal["eligible", "insufficient_data"]


class MedicalRecordExtractionDTO(BaseModel):
    id: str
    version: int
    status: AsyncJobStatus
    error_summary: str | None = None
    document_type: str | None = None
    summary_text: str | None = None
    raw_text: str | None = None
    structured_payload: dict[str, Any] | None = None
    source_model: str | None = None
    prompt_version: str | None = None
    completed_at: datetime | None = None
    updated_at: datetime


class MedicalRecordFileDTO(BaseModel):
    id: str
    archive_id: str
    original_filename: str
    content_type: str
    size_bytes: int
    processing_status: AsyncJobStatus
    processing_error: str | None = None
    created_at: datetime
    updated_at: datetime
    processed_at: datetime | None = None
    latest_extraction: MedicalRecordExtractionDTO | None = None


class MedicalRecordArchiveSummaryDTO(BaseModel):
    id: str
    title: str
    description: str | None = None
    created_at: datetime
    updated_at: datetime
    file_count: int
    report_count: int
    latest_activity_at: datetime | None = None
    latest_report: "MedicalRecordReportSummaryDTO | None" = None


class MedicalRecordArchiveDetailDTO(MedicalRecordArchiveSummaryDTO):
    disclaimer: str
    consent_policy: str
    retention_policy: str
    delete_policy: str
    export_policy: str
    files: list[MedicalRecordFileDTO] = Field(default_factory=list)
    reports: list["MedicalRecordReportSummaryDTO"] = Field(default_factory=list)


class MedicalRecordArchiveListResponse(BaseModel):
    archives: list[MedicalRecordArchiveSummaryDTO]


class CreateMedicalRecordArchiveRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=4000)


class CreateMedicalRecordArchiveResponse(BaseModel):
    archive: MedicalRecordArchiveDetailDTO


class UploadMedicalRecordFileResponse(BaseModel):
    file: MedicalRecordFileDTO


class MedicalRecordFileListResponse(BaseModel):
    files: list[MedicalRecordFileDTO]


class CreateMedicalRecordReportRequest(BaseModel):
    report_window_days: int = Field(default=30, ge=7, le=365)
    monitoring_window_days: int = Field(default=30, ge=7, le=365)
    medication_window_days: int = Field(default=30, ge=7, le=365)


class MedicalRecordReportSummaryDTO(BaseModel):
    id: str
    archive_id: str
    archive_title: str | None = None
    version: int
    title: str
    status: AsyncJobStatus
    pdf_status: AsyncJobStatus
    error_summary: str | None = None
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None
    generated_at: datetime | None = None
    summary: str | None = None
    pdf_ready: bool
    pdf_file_name: str | None = None
    report_window_label: str | None = None


class MedicalRecordReportDetailDTO(MedicalRecordReportSummaryDTO):
    archive_description: str | None = None
    disclaimer: str
    disclaimer_version: str
    model_name: str | None = None
    prompt_version: str | None = None
    report_window_start: date
    report_window_end: date
    monitoring_window_start: date
    monitoring_window_end: date
    medication_window_start: date
    medication_window_end: date
    input_snapshot: dict[str, Any] | None = None
    report_payload: dict[str, Any] | None = None
    narrative_text: str | None = None
    has_pdf: bool
    sections: list[dict[str, str]] = Field(default_factory=list)
    source_files: list[MedicalRecordFileDTO] = Field(default_factory=list)
    history: list[MedicalRecordReportSummaryDTO] = Field(default_factory=list)


class MedicalRecordReportListResponse(BaseModel):
    reports: list[MedicalRecordReportSummaryDTO]


class CreateMedicalRecordReportResponse(BaseModel):
    report: MedicalRecordReportSummaryDTO


class RehabPlanItemDTO(BaseModel):
    template_id: str
    name: str
    category: str
    duration_minutes: int
    frequency_label: str
    cautions: list[str] = Field(default_factory=list)


class RehabPlanDTO(BaseModel):
    id: str
    title: str
    status: RehabPlanStatus
    scenario: str
    version: int
    rationale: str
    items: list[RehabPlanItemDTO]
    risk_flags: list[str] = Field(default_factory=list)
    requires_confirmation: bool
    difference_summary: str | None = None
    generated_at: datetime
    confirmed_at: datetime | None = None
    activated_at: datetime | None = None


class RehabEvidenceSummaryDTO(BaseModel):
    as_of_date: date
    evaluation_window: Literal["calendar_day"]
    medication_window_summary: str
    tremor_trend_summary: str
    signal_consistency: RehabConflictStatus
    explanation: str
    generation_eligibility: RehabGenerationEligibility
    missing_inputs: list[str] = Field(default_factory=list)


class RehabGuidanceResponse(BaseModel):
    active_plan: RehabPlanDTO | None = None
    candidate_plan: RehabPlanDTO | None = None
    evidence_summary: RehabEvidenceSummaryDTO
    conflict_status: RehabConflictStatus
    disclaimer: str
    generated_at: datetime | None = None


class GenerateRehabGuidanceRequest(BaseModel):
    as_of_date: date


MedicalRecordArchiveSummaryDTO.model_rebuild()
MedicalRecordArchiveDetailDTO.model_rebuild()
MedicalRecordReportDetailDTO.model_rebuild()

File: tremor-guard-frontend/src/types/domain.ts
export type NavKey =
  | 'overview'
  | 'ai-doctor'
  | 'medication'
  | 'rehab-guidance'
  | 'medical-records'
  | 'reports'
  | 'profile'

export interface PatientProfile {
  id: string
  name: string
  age: number
  gender: string
  diagnosis: string
  duration: string
  hospital: string
  deviceId: string
}

export interface DeviceStatus {
  battery: number
  connection: 'stable' | 'syncing' | 'offline'
  connectionLabel: string
  lastSync: string
  availableDays: string
  firmware: string
}

export interface TremorMetricSummary {
  label: string
  value: string | number
  unit: string
  subtitle: string
  tone: 'neutral' | 'safe' | 'alert'
}

export interface TremorTrendPoint {
  time: string
  amplitude: number
  label?: string
  medicationTaken?: boolean
}

export interface AiInsight {
  id: string
  title: string
  summary: string
  emphasis?: string
}

export interface OverviewEvidenceReadiness {
  hasDeviceBinding: boolean
  hasMonitoringEvents: boolean
  monitoringEventCount: number
  hasMedicationLogs: boolean
  medicationLogCount: number
  hasMedicalRecordArchives: boolean
  medicalRecordArchiveCount: number
  aiInterpretationReady: boolean
  rehabPlanReady: boolean
  healthReportReady: boolean
  nextSteps: string[]
}

export type AiChatActionKind =
  | 'confirm_plan'
  | 'view_plan_detail'
  | 'download_plan_pdf'
  | 'view_report_online'
  | 'download_report_pdf'

export interface AiChatAction {
  key: string
  label: string
  kind: AiChatActionKind
  apiPath?: string | null
  url?: string | null
  downloadName?: string | null
}

export interface AiChatActionCard {
  type: 'rehab_plan_candidate' | 'health_report_candidate'
  agentType?: string | null
  title: string
  summary: string
  status: string
  resourceId: string
  resourcePath?: string | null
  pipelineState?: MedicalRecordReportPipelineState | null
  actions: AiChatAction[]
}

export interface ChatMessage {
  id: string | number
  role: 'assistant' | 'user'
  content: string
  actionCards?: AiChatActionCard[]
}

export interface AiChatUsage {
  promptTokens?: number | null
  completionTokens?: number | null
  totalTokens?: number | null
}

export interface MedicationEntry {
  id: number
  time: string
  name: string
  dose: string
  status: 'taken' | 'pending' | 'missed'
}

export type RehabGenerationEligibility = 'eligible' | 'insufficient_data'

export type RehabConflictStatus = 'consistent' | 'conflicting' | 'insufficient_data' | 'unknown'

export type RehabPlanStatus =
  | 'active_only'
  | 'candidate_pending_confirmation'
  | 'candidate_confirmed'
  | 'candidate_superseded'

export interface RehabPlanItem {
  templateId: string
  name: string
  category: string
  durationMinutes: number
  frequencyLabel: string
  cautions: string[]
  goal?: string | null
  preparation: string[]
  steps: string[]
  completionCheck?: string | null
}

export interface RehabPlan {
  id: string
  title: string
  status: RehabPlanStatus
  scenario: string
  version: number
  rationale: string
  items: RehabPlanItem[]
  riskFlags: string[]
  requiresConfirmation: boolean
  generatedAt?: string | null
  confirmedAt?: string | null
  activatedAt?: string | null
}

export interface RehabEvidenceSummary {
  asOfDate?: string | null
  evaluationWindow?: string | null
  medicationWindowSummary: string
  tremorTrendSummary: string
  signalConsistency: string
  explanation: string
}

export interface RehabGuidanceViewData {
  activePlan: RehabPlan | null
  candidatePlan: RehabPlan | null
  evidenceSummary: RehabEvidenceSummary
  conflictStatus: RehabConflictStatus
  disclaimer: string
  generatedAt?: string | null
  generationEligibility: RehabGenerationEligibility
  missingInputs: string[]
}

export interface ReportSummary {
  id: string
  date: string
  type: string
  size: string
  status?: string
  kind?: 'legacy_monitoring_summary'
}

export type LegacyMonitoringReportSummary = ReportSummary

export type MedicalRecordProcessingStatus = 'queued' | 'processing' | 'succeeded' | 'failed'

export interface MedicalRecordPipelineStage {
  status: MedicalRecordProcessingStatus
  startedAt?: string | null
  completedAt?: string | null
  detail?: string | null
  error?: string | null
}

export interface MedicalRecordReportPipelineState {
  template: MedicalRecordPipelineStage
  llm: MedicalRecordPipelineStage
  pdf: MedicalRecordPipelineStage
  updatedAt?: string | null
}

export interface MedicalRecordArchiveSummary {
  id: string
  title: string
  description?: string | null
  createdAt: string
  updatedAt: string
  fileCount: number
  reportCount: number
  latestActivityAt?: string | null
  latestReportId?: string | null
  latestReportVersion?: number | null
  latestReportStatus?: MedicalRecordProcessingStatus | null
}

export interface MedicalRecordExtraction {
  status: MedicalRecordProcessingStatus
  summary?: string | null
  highlights: string[]
  extractedAt?: string | null
  errorSummary?: string | null
}

export interface MedicalRecordFile {
  id: string
  archiveId: string
  fileName: string
  fileType: string
  sizeLabel: string
  uploadedAt: string
  processingStatus: MedicalRecordProcessingStatus
  previewUrl?: string | null
  statusSummary?: string | null
  extraction?: MedicalRecordExtraction | null
}

export interface MedicalRecordReportSummary {
  id: string
  agentType?: string | null
  archiveId: string
  archiveTitle: string
  version: number
  title: string
  status: MedicalRecordProcessingStatus
  generatedAt?: string | null
  summary?: string | null
  pdfReady: boolean
  pdfFileName?: string | null
  reportWindowLabel?: string | null
  templateName?: string | null
  templateVersion?: string | null
  pipelineState?: MedicalRecordReportPipelineState | null
  qualityWarnings: string[]
}

export interface MedicalRecordReportSection {
  id: string
  title: string
  body: string
}

export interface MedicalRecordInputSnapshot {
  reportWindow?: string | null
  monitoringWindow?: string | null
  medicationWindow?: string | null
  disclaimerVersion?: string | null
  promptVersion?: string | null
  modelVersion?: string | null
}

export interface MedicalRecordArchiveDetail extends MedicalRecordArchiveSummary {
  disclaimer: string
  files: MedicalRecordFile[]
  reports: MedicalRecordReportSummary[]
}

export interface MedicalRecordReportDetail extends MedicalRecordReportSummary {
  disclaimer: string
  archiveDescription?: string | null
  reportMarkdown?: string | null
  sections: MedicalRecordReportSection[]
  sourceFiles: MedicalRecordFile[]
  history: MedicalRecordReportSummary[]
  inputSnapshot?: MedicalRecordInputSnapshot | null
}

export type HealthReportSummary = MedicalRecordReportSummary
export type HealthReportDetail = MedicalRecordReportDetail

export interface ConsentSettings {
  shareWithDoctor: boolean
  ragAnalysisEnabled: boolean
  cloudSyncEnabled: boolean
}

export interface CurrentUser {
  id: string
  email: string
  displayName: string
  status: string
  onboardingState: 'profile_required' | 'device_binding_required' | 'active'
  lastLoginAt?: string | null
  hasProfile: boolean
  hasBoundDevice: boolean
  boundDeviceSerial?: string | null
}

export interface AuthSession {
  accessToken: string
  refreshToken: string
  userId: string
  accessTokenExpiresAt: string
  refreshTokenExpiresAt: string
}

export interface ProfileCompletionStatus {
  onboardingState: CurrentUser['onboardingState']
  hasProfile: boolean
  hasBoundDevice: boolean
}

export interface DeviceBinding {
  id: string
  deviceSerial: string
  deviceName: string
  firmwareVersion: string
  bindingStatus: string
  boundAt?: string | null
  unboundAt?: string | null
}
